lib/auth/auth.go

```diff
@@ -1214,6 +1225,92 @@ func (a *Server) DeleteMFADeviceWithToken(ctx context.Context, req *proto.Delete
 	return nil
 }
 
+// AddMFADeviceWithToken adds a mfa device for the user defined in token.
+func (a *Server) AddMFADeviceWithToken(ctx context.Context, req *proto.AddMFADeviceWithTokenRequest) error {
+	token, err := a.GetUserToken(ctx, req.GetTokenID())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Restrict to certain kinds of user token.
+	if token.GetSubKind() != UserTokenTypePrivilege {
+		return trace.BadParameter("invalid token")
+	}
+
+	if token.Expiry().Before(a.clock.Now().UTC()) {
+		return trace.BadParameter("expired token")
+	}
+
+	authPref, err := a.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	secondFactor := authPref.GetSecondFactor()
+	var newDevice *types.MFADevice
+
+	switch req.GetSecondFactorCred().(type) {
+	case *proto.AddMFADeviceWithTokenRequest_U2FRegisterResponse:
+		if secondFactor == constants.SecondFactorOTP {
+			return trace.BadParameter("cluster only allows OTP for second factor")
+		}
+
+		cfg, err := authPref.GetU2F()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		device, err := a.createNewU2FDevice(ctx, newU2FDeviceRequest{
```

Comment:
Device creation only happened at addMFADeviceRegisterChallenge - I don't think we should be handling this in 2 places. This is only getting more complex with the addition of Webauthn.